### PR TITLE
Yield an error if no polkit authentication agent was found

### PR DESCRIPTION
**Pull request by jviotti**
`pkexec` prints an error to `stderr` if no authentication agent was
found, and if we pass the `--disable-internal-agent` option.

Currently, this module doesn't handle this edge case by yielding a
meaningful error, and therefore causes client applications to usually
miss it.

Fixes: https://github.com/jorangreef/sudo-prompt/issues/28
Signed-off-by: Juan Cruz Viotti <jviottidc@gmail.com>

### CODE REVIEW

**Comment by jorangreef** [comment]
Thanks @jviotti, this looks good.

There may be an issue with PolicyKit giving internationalized error messages on some platforms, in which case our `stderr` check will not translate. We have had this kind of thing in the past on Asian machines, but let's try this and see if we run into it again. If our check gets an internationalized `stderr`, then hopefully the user will be able to make sense of it anyway.

What is the value of `error` if `stderr` contains "No authentication agent found"?
Perhaps we could make the `stderr` check more restricted, i.e. only do the check if `error` is as we expect?

I think the `PERMISSION_DENIED` and `NO_POLKIT_AGENT` blocks should be mutually exclusive, i.e. we should use an `if`...`else`? That way, we don't open up the possibility of re-assigning `error` more than once in rare cases.

We also need to double-check that we don't fire a `PERMISSION_DENIED` when we should fire a `NO_POLKIT_AGENT`.

Would you also please remove the leading and trailing newlines?

Thanks!

**Comment by jviotti** [comment]
Hi @jorangreef ,

> There may be an issue with PolicyKit giving internationalized error messages on some platforms, in which case our `stderr` check will not translate.

As far as I can see, the error we're checking for here is hardcoded into `pkexec`: https://cgit.freedesktop.org/polkit/tree/src/programs/pkexec.c#n866, but maybe `pkexec` is being patched in some Linux systems for internationalisation purposes (I doubt so)?

> What is the value of error if stderr contains "No authentication agent found"?
Perhaps we could make the stderr check more restricted, i.e. only do the check if error is as we expect?

The user gets the classic `PERMISSION_DENIED` error. I'm promisifying this module in Etcher, which means that the `.catch()` block is instantly called without allowing me to check `stderr` at all.

I'll update the PR as per your suggestions in a second.

File: index.js [hunk 1]
```diff
@@ -127,7 +127,9 @@ function Linux(instance, end) {
       command = command.join(' ');
       Node.child.exec(command,
         function(error, stdout, stderr) {
-          if (error && /Request dismissed|Command failed/i.test(error)) {
+          if (/No authentication agent found/.test(stderr)) {
```
Comment (jorangreef) [inline index.js]: Could you add an `error &&` to the start of the condition? This restricts the check to our expectations, and prevents us from scanning `stderr` (which may be large) unnecessarily.

Alternatively, we could also nest the `if...else` block within a `if (error) {` block.

**Comment by jorangreef** [comment]
Thanks for the link to the polkit source. Added one comment on the PR source.

File: index.js (hunk 1 above)
Comment (jviotti) [inline index.js]: Sure, updated!

**Pull request merged**